File: output/GPT4o-Mini/xCodeEval_4000/C++/Java/7b63345be437e04c0f3b18ccf1e0ee68.java
import java.util.*;
import java.io.*;

class DSum {
    static class Solver {
        BufferedReader in;
        PrintWriter out;
        int N, K;
        List<List<Long>> A;
        long[] A_sum;
        long best;

        Solver(BufferedReader in, PrintWriter out) {
            this.in = in;
            this.out = out;
        }

        void rd() throws IOException {
            StringTokenizer st = new StringTokenizer(in.readLine());
            N = Integer.parseInt(st.nextToken());
            K = Integer.parseInt(st.nextToken());
            A = new ArrayList<>(N);
            A_sum = new long[N];

            for (int i = 0; i < N; i++) {
                int t = Integer.parseInt(in.readLine());
                List<Long> temp = new ArrayList<>();
                StringTokenizer st2 = new StringTokenizer(in.readLine());
                for (int j = 0; j < t; j++) {
                    temp.add(Long.parseLong(st2.nextToken()));
                }
                if (t > K) {
                    temp = temp.subList(0, K);
                }
                A.add(temp);
                A_sum[i] = temp.stream().mapToLong(Long::longValue).sum();
            }
        }

        void add_item(long[] knapsack, int size, long sum) {
            for (int k = K - size; k >= 0; k--) {
                knapsack[k + size] = Math.max(knapsack[k + size], knapsack[k] + sum);
            }
        }

        void recurse(int start, int end, long[] knapsack) {
            if (start >= end) return;

            if (end - start == 1) {
                long sum = 0;
                best = Math.max(best, knapsack[K]);

                for (int prefix = 1; prefix <= A.get(start).size(); prefix++) {
                    sum += A.get(start).get(prefix - 1);
                    best = Math.max(best, sum + knapsack[K - prefix]);
                }
                return;
            }

            int mid = (start + end) / 2;
            long[] state = Arrays.copyOf(knapsack, knapsack.length);

            for (int i = start; i < mid; i++) {
                add_item(state, A.get(i).size(), A_sum[i]);
            }

            recurse(mid, end, state);
            state = Arrays.copyOf(knapsack, knapsack.length);

            for (int i = mid; i < end; i++) {
                add_item(state, A.get(i).size(), A_sum[i]);
            }

            recurse(start, mid, state);
        }

        void solve() throws IOException {
            rd();
            best = 0;
            long[] knapsack = new long[K + 1];
            recurse(0, N, knapsack);
            out.println(best);
        }
    }

    static void solve(BufferedReader in, PrintWriter out) throws IOException {
        new Solver(in, out).solve();
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        PrintWriter out = new PrintWriter(System.out);
        solve(in, out);
        out.close();
    }
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/7b63345be437e04c0f3b18ccf1e0ee68.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <stdint.h>
#include <string.h>

#define MAX_N 100000
#define MAX_K 100

int64_t max(int64_t a, int64_t b) {
    return (a > b) ? a : b;
}

void add_item(int64_t *knapsack, int size, int64_t sum, int K) {
    for (int k = K - size; k >= 0; k--) {
        knapsack[k + size] = max(knapsack[k + size], knapsack[k] + sum);
    }
}

void recurse(int start, int end, int64_t *knapsack, int K, int N, int64_t **A, int64_t *A_sum, int64_t *best) {
    if (start >= end) {
        return;
    }

    if (end - start == 1) {
        int64_t sum = 0;
        *best = max(*best, knapsack[K]);

        for (int prefix = 1; prefix <= A[start][0]; prefix++) {
            sum += A[start][prefix];
            *best = max(*best, sum + knapsack[K - prefix]);
        }

        return;
    }

    int mid = (start + end) / 2;
    int64_t *state = (int64_t *)malloc((K + 1) * sizeof(int64_t));
    memcpy(state, knapsack, (K + 1) * sizeof(int64_t));

    for (int i = start; i < mid; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(mid, end, state, K, N, A, A_sum, best);
    free(state);

    state = (int64_t *)malloc((K + 1) * sizeof(int64_t));
    memcpy(state, knapsack, (K + 1) * sizeof(int64_t));

    for (int i = mid; i < end; i++) {
        add_item(state, A[i][0], A_sum[i], K);
    }

    recurse(start, mid, state, K, N, A, A_sum, best);
    free(state);
}

void solve() {
    int N, K;
    scanf("%d %d", &N, &K);

    int64_t **A = (int64_t **)malloc(N * sizeof(int64_t *));
    int64_t *A_sum = (int64_t *)malloc(N * sizeof(int64_t));

    for (int i = 0; i < N; i++) {
        int t;
        scanf("%d", &t);
        A[i] = (int64_t *)malloc((t + 1) * sizeof(int64_t));
        A[i][0] = t;

        for (int j = 1; j <= t; j++) {
            scanf("%lld", &A[i][j]);
        }

        if (t > K) {
            A[i][0] = K;
        }

        A_sum[i] = 0;

        for (int j = 1; j <= A[i][0]; j++) {
            A_sum[i] += A[i][j];
        }
    }

    int64_t *knapsack = (int64_t *)malloc((K + 1) * sizeof(int64_t));
    memset(knapsack, 0, (K + 1) * sizeof(int64_t));

    int64_t best = 0;
    recurse(0, N, knapsack, K, N, A, A_sum, &best);

    printf("%lld\n", best);

    for (int i = 0; i < N; i++) {
        free(A[i]);
    }

    free(A);
    free(A_sum);
    free(knapsack);
}

int main() {
    solve();
    return 0;
}
